generate/src/contract/events.rs

```diff
@@ -1,21 +1,149 @@
-use crate::contract::{types, Context};
+use crate::contract::{methods, types, Context};
 use crate::util;
 use anyhow::Result;
 use ethcontract_common::abi::{Event, Hash};
+use ethcontract_common::abiext::EventExt;
 use inflector::Inflector;
 use proc_macro2::{Literal, TokenStream};
 use quote::quote;
 
 pub(crate) fn expand(cx: &Context) -> Result<TokenStream> {
+    let structs_mod = expand_structs_mod(cx)?;
     let filters = expand_filters(cx)?;
     let all_events = expand_all_events(cx);
 
     Ok(quote! {
+        #structs_mod
         #filters
         #all_events
     })
 }
 
+fn expand_structs_mod(cx: &Context) -> Result<TokenStream> {
+    let structs = cx
+        .artifact
+        .abi
+        .events()
+        .map(|event| expand_struct(event))
+        .collect::<Result<Vec<_>>>()?;
+    if structs.is_empty() {
+        return Ok(quote! {});
+    }
+    Ok(quote! {
+        /// Module containing all generated data models for this contract's
+        /// events.
+        pub mod events {
+            use super::ethcontract;
+
+            #( #structs )*
+        }
+    })
+}
+
+fn expand_struct(event: &Event) -> Result<TokenStream> {
+    let event_name = expand_struct_name(event);
+
+    let signature = expand_hash(event.signature());
+    let abi_signature = Literal::string(&event.abi_signature());
+
+    let params = event
+        .inputs
+        .iter()
+        .enumerate()
+        .map(|(i, input)| {
+            // NOTE: Events can contain nameless values.
+            let name = methods::expand_input_name(i, &input.name);
+            let ty = types::expand(&input.kind)?;
+            Ok((name, ty))
+        })
+        .collect::<Result<Vec<_>>>()?;
+
+    let param_names = params
+        .iter()
+        .map(|(name, _)| name)
+        .cloned()
+        .collect::<Vec<_>>();
+    let (params_def, params_cstr) = if event.inputs.iter().all(|input| input.name.is_empty()) {
+        let fields = params
+            .iter()
+            .map(|(_, ty)| quote! { pub #ty, })
+            .collect::<Vec<_>>();
+
+        let defs = quote! { ( #( #fields )* ) };
+        let cstr = quote! { ( #( #param_names ),* ) };
+
+        (defs, cstr)
+    } else {
+        let fields = params
+            .iter()
+            .map(|(name, ty)| quote! { pub #name: #ty, })
+            .collect::<Vec<_>>();
+
+        let defs = quote! { { #( #fields )* } };
+        let cstr = quote! { { #( #param_names ),* } };
+
+        (defs, cstr)
+    };
+
+    let param_tokens = params
+        .iter()
+        .map(|(name, ty)| {
+            quote! {
+                let #name = #ty::from_token(tokens.next().unwrap())?;
+            }
+        })
+        .collect::<Vec<_>>();
+    let param_len = Literal::usize_unsuffixed(params.len());
```

Comment:
This part is hard to follow for me. Could we split it up into smaller functions and/or add tests? Don't do it if you think it's fine.

Reply:
@e00E let me know if its better now.

Reply:
Yes, thanks.